### PR TITLE
Added CASE Authenticated Tag (CAT) Support To the Commissioner APIs.

### PR DESCRIPTION
**Pull request by emargolis**
#### Problem
Current code doesn't support CASE Authenticated Tags in the Operation Certificates generation logic.

#### Change overview
Added CAT support

#### Testing
existing tests

### CODE REVIEW

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -1227,6 +1227,8 @@ CHIP_ERROR DeviceCommissioner::ProcessCSR(DeviceProxy * proxy, const ByteSpan &
         mOperationalCredentialsDelegate->SetFabricIdForNextNOCRequest(mFabricInfo->GetFabricId());
     }
 
+    mOperationalCredentialsDelegate->SetCATForNextNOCRequest(kTestDeviceCAT);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1230]: Wait, why would we do this unconditionally, even in normal operation?  This does not look right to me...

File: src/controller/ExampleOperationalCredentialsIssuer.cpp [hunk 2]
```diff
@@ -106,7 +106,7 @@ CHIP_ERROR ExampleOperationalCredentialsIssuer::Initialize(PersistentStorageDele
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR ExampleOperationalCredentialsIssuer::GenerateNOCChainAfterValidation(NodeId nodeId, FabricId fabricId,
+CHIP_ERROR ExampleOperationalCredentialsIssuer::GenerateNOCChainAfterValidation(NodeId nodeId, FabricId fabricId, CASEAuthTag cat,
```
Comment (bzbarsky-apple) [inline src/controller/ExampleOperationalCredentialsIssuer.cpp:109]: Why only one CAT?

File: src/controller/OperationalCredentialsDelegate.h [hunk 3]
```diff
@@ -78,6 +79,13 @@ class DLL_EXPORT OperationalCredentialsDelegate
      */
     virtual void SetFabricIdForNextNOCRequest(FabricId fabricId) {}
 
+    /**
+     *   This function sets the CASE Authenticated Tag (CAT) for which the next NOC Chain should be generated.
+     *   This API is not required to be implemented if the delegate implementation has other mechanisms
+     *   to find the CAT.
+     */
+    virtual void SetCATForNextNOCRequest(CASEAuthTag cat) {}
```
Comment (bzbarsky-apple) [inline src/controller/OperationalCredentialsDelegate.h:87]: Why only one, a opposed to the list of CATs to use?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/controller/CHIPDeviceController.cpp (hunk 1 above)
Comment (msandstedt) [reply src/controller/CHIPDeviceController.cpp:1230]: Agree.  This type of thing, if it's going to be done, needs to be done from chip-tool.

**Review by msandstedt — CHANGES_REQUESTED** [review]
This doesn't really make sense.  The OperationalCredentialsDelegate interface (shared by all users of `chip::Controller::DeviceCommissioner` cannot actually be designed to anticipate the types of context injection need to happen for apps.

Given this, why isn't this just added to the ExampleOperationalCredentialsDelegate?

File: src/controller/ExampleOperationalCredentialsIssuer.cpp (hunk 2 above)
Comment (emargolis) [reply src/controller/ExampleOperationalCredentialsIssuer.cpp]: Fixed that. Changed that to the list of CATs (CATValues).

File: src/controller/CHIPDeviceController.cpp (hunk 1 above)
Comment (emargolis) [reply src/controller/CHIPDeviceController.cpp]: Yes, thanks for suggestion. I agree and moved it to the application layer.

**Comment by emargolis** [comment]
> This doesn't really make sense. The OperationalCredentialsDelegate interface (shared by all users of `chip::Controller::DeviceCommissioner` cannot actually be designed to anticipate the types of context injection need to happen for apps.
> 
> Given this, why isn't this just added to the ExampleOperationalCredentialsDelegate?

This was addressed

File: examples/chip-tool/commands/common/CredentialIssuerCommands.h [hunk 4]
```diff
@@ -60,14 +60,15 @@ class CredentialIssuerCommands
...
      * @param[inout] rcac  Buffer to hold the Root Certificate of the generated NOC Chain.
      * @param[inout] icac  Buffer to hold the Intermediate Certificate of the generated NOC Chain.
      * @param[inout] noc   Buffer to hold the Leaf Certificate of the generated NOC Chain.
      *
      * @return CHIP_ERROR CHIP_NO_ERROR on success, or corresponding error code.
      */
-    virtual CHIP_ERROR GenerateControllerNOCChain(chip::NodeId nodeId, chip::FabricId fabricId, chip::Crypto::P256Keypair & keypair,
-                                                  chip::MutableByteSpan & rcac, chip::MutableByteSpan & icac,
-                                                  chip::MutableByteSpan & noc) = 0;
+    virtual CHIP_ERROR GenerateControllerNOCChain(chip::NodeId nodeId, chip::FabricId fabricId, chip::CATValues cats,
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/CredentialIssuerCommands.h:71]: Should that be a `const CATValues & cats`, so we don't end up copying it as we go?

File: examples/chip-tool/config/PersistentStorage.cpp [hunk 5]
```diff
@@ -236,3 +237,29 @@ CHIP_ERROR PersistentStorage::SetLocalNodeId(NodeId value)
...
+        err = cats.Deserialize(serializedCATs);
+        if (err == CHIP_NO_ERROR)
+        {
+            return cats;
+        }
+    }
+    return kTestControllerCATs;
+}
+
+CHIP_ERROR PersistentStorage::SetLocalCATs(CATValues cats)
```
Comment (bzbarsky-apple) [inline examples/chip-tool/config/PersistentStorage.cpp:259]: Please pass by const reference?

File: src/credentials/CHIPCert.cpp [hunk 6]
```diff
@@ -577,6 +577,19 @@ CHIP_ERROR ChipDN::AddAttribute(chip::ASN1::OID oid, uint64_t val)
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR ChipDN::AddCATs(chip::CATValues cats)
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCert.cpp:580]: Again, should probably pass by const reference.

File: src/darwin/Framework/CHIP/CHIPDeviceController.mm [hunk 7]
```diff
@@ -217,8 +217,8 @@ - (BOOL)startup:(_Nullable id<CHIPPersistentStorageDelegate>)storageDelegate
 
         chip::MutableByteSpan icac;
 
-        errorCode = _operationalCredentialsDelegate->GenerateNOCChainAfterValidation(
-            _localDeviceId, 0, ephemeralKey.Pubkey(), rcac, icac, noc);
+        errorCode = _operationalCredentialsDelegate->GenerateNOCChainAfterValidation(_localDeviceId, 0,
+            { { chip::kTestControllerCAT, chip::kUndefinedCAT, chip::kUndefinedCAT } }, ephemeralKey.Pubkey(), rcac, icac, noc);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPDeviceController.mm:221]: No, wait, why would we do that, exactly?  This is not test code; it should not be doing this.

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm [hunk 8]
```diff
@@ -203,11 +204,12 @@ static BOOL isRunningTests(void)
     }
 
     ChipDN noc_dn;
-    noc_dn.AddAttribute(chip::ASN1::kOID_AttributeType_ChipFabricId, fabricId);
-    noc_dn.AddAttribute(chip::ASN1::kOID_AttributeType_ChipNodeId, nodeId);
+    ReturnErrorOnFailure(noc_dn.AddAttribute(chip::ASN1::kOID_AttributeType_ChipFabricId, fabricId));
+    ReturnErrorOnFailure(noc_dn.AddAttribute(chip::ASN1::kOID_AttributeType_ChipNodeId, nodeId));
+    ReturnErrorOnFailure(noc_dn.AddCATs(cats));
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:209]: No, we don't want to be doing this.  This information would need to be coming from the actual application code or trust store, depending, not from some sort of hardcoded test values.

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm [hunk 9]
```diff
@@ -237,6 +239,27 @@ static BOOL isRunningTests(void)
...
 
+static CHIP_ERROR AssignCATsForGivenDAC(const chip::ByteSpan & DAC, chip::CATValues & assignedCATs)
+{
+    uint16_t dacVid = VendorId::NotSpecified;
+    uint16_t dacPid = 0;
+
+    ReturnErrorOnFailure(ExtractDNAttributeFromX509Cert(MatterOid::kVendorId, DAC, dacVid));
+    ReturnErrorOnFailure(ExtractDNAttributeFromX509Cert(MatterOid::kProductId, DAC, dacPid));
+
+    if (IsTestVendorId(dacVid)) {
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:250]: I am entirely unconvinced we want this either.  Why do we want it here?

File: src/lib/core/CASEAuthTag.h [hunk 10]
```diff
@@ -56,6 +57,30 @@ struct CATValues
...
+    {
+        for (size_t i = 0; i < kMaxSubjectCATAttributeCount; i++)
+        {
+            uint32_t littleEndianCAT = Encoding::LittleEndian::HostSwap32(values[i]);
+            memcpy(&outSerialized[i * sizeof(littleEndianCAT)], &littleEndianCAT, sizeof(littleEndianCAT));
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Deserialize(const Serialized inSerialized)
```
Comment (bzbarsky-apple) [inline src/lib/core/CASEAuthTag.h:74]:
```suggestion
    CHIP_ERROR Deserialize(const Serialized & inSerialized)
```

File: src/lib/core/CASEAuthTag.h [hunk 11]
```diff
@@ -56,6 +57,30 @@ struct CATValues
...
+    {
+        for (size_t i = 0; i < kMaxSubjectCATAttributeCount; i++)
+        {
+            uint32_t littleEndianCAT;
+            memcpy(&littleEndianCAT, &inSerialized[i * sizeof(littleEndianCAT)], sizeof(littleEndianCAT));
+            values[i] = Encoding::LittleEndian::HostSwap32(littleEndianCAT);
```
Comment (bzbarsky-apple) [inline src/lib/core/CASEAuthTag.h:80]: Why not use Get32/Read32, without needing to memcpy?

File: src/lib/core/CHIPVendorIdentifiers.hpp [hunk 12]
```diff
@@ -50,4 +50,9 @@ enum VendorId : uint16_t
     NotSpecified = 0xFFFFu
 };
 
+constexpr bool IsTestVendorId(uint16_t vid)
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPVendorIdentifiers.hpp:53]: Shouldn't this take VendorId?

File: src/lib/core/CASEAuthTag.h (hunk 11 above)
Comment (emargolis) [reply src/lib/core/CASEAuthTag.h]: updated

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm (hunk 9 above)
Comment (emargolis) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm]: The implementation was changed

File: examples/chip-tool/config/PersistentStorage.cpp (hunk 5 above)
Comment (emargolis) [reply examples/chip-tool/config/PersistentStorage.cpp]: done

File: examples/chip-tool/commands/common/CredentialIssuerCommands.h (hunk 4 above)
Comment (emargolis) [reply examples/chip-tool/commands/common/CredentialIssuerCommands.h]: done

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm (hunk 8 above)
Comment (emargolis) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:231]: Correct, this `cats` information propagates from the application code now.

File: src/darwin/Framework/CHIP/CHIPDeviceController.mm (hunk 7 above)
Comment (emargolis) [reply src/darwin/Framework/CHIP/CHIPDeviceController.mm]: I've updated the implementation. Reading CATs Configuration from the persistent memory. However, the exact implementation of the mechanism is TODO - subject for another PR.

File: src/darwin/Framework/CHIP/CHIPDeviceController.mm [hunk 13]
```diff
@@ -272,6 +277,32 @@ - (NSNumber *)_getControllerNodeId
     return [NSNumber numberWithUnsignedLongLong:_localDeviceId];
 }
 
+- (CATValues)getControllerCATs
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPDeviceController.mm:280]: Why does this function exist?  It's not used....

File: src/darwin/Framework/CHIP/CHIPDeviceController.mm [hunk 14]
```diff
@@ -218,7 +223,7 @@ - (BOOL)startup:(_Nullable id<CHIPPersistentStorageDelegate>)storageDelegate
         chip::MutableByteSpan icac;
 
         errorCode = _operationalCredentialsDelegate->GenerateNOCChainAfterValidation(
-            _localDeviceId, 0, ephemeralKey.Pubkey(), rcac, icac, noc);
+            _localDeviceId, 0, _localDeviceCATs, ephemeralKey.Pubkey(), rcac, icac, noc);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPDeviceController.mm:226]: Again, this CATs stuff should be coming from the consumer of the framework, not from this code....

Can this code just pass "no cats" to `GenerateNOCChainAfterValidation` and not add these bits that I'm pretty sure we don't actually want inside the framework?

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm [hunk 15]
```diff
@@ -100,6 +101,24 @@ static BOOL isRunningTests(void)
...
+    char catsTableString[kDeviceCATsTableMaxLength];
+    uint16_t catsTableStringLen = sizeof(catsTableString);
+    CHIP_ERROR err = mStorage->SyncGetKeyValue(CHIP_COMMISSIONER_DEVICE_CATS_KEY, catsTableString, catsTableStringLen);
+    if (err == CHIP_NO_ERROR) {
+        // TODO: Implement CATs table structure and logic that extracts CATs from that table
+        // for a given VID and PID values.
+        cats = chip::kUndefinedCATs;
+    } else {
+        cats = chip::kUndefinedCATs;
+    }
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:118]:
```suggestion
    cats = chip::kUndefinedCATs;
```

File: src/protocols/secure_channel/PASESession.h [hunk 16]
```diff
@@ -91,8 +91,8 @@ struct PASEVerifier
     uint8_t mW0[kP256_FE_Length];
     uint8_t mL[kP256_Point_Length];
 
-    CHIP_ERROR Serialize(MutableByteSpan & outSerialized);
-    CHIP_ERROR Deserialize(ByteSpan inSerialized);
+    CHIP_ERROR Serialize(MutableByteSpan & outSerialized) const;
+    CHIP_ERROR Deserialize(const ByteSpan inSerialized);
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/PASESession.h:95]:
```suggestion
    CHIP_ERROR Deserialize(const ByteSpan & inSerialized);
```

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm [hunk 17]
```diff
@@ -272,6 +293,14 @@ static BOOL isRunningTests(void)
     chip::Crypto::P256PublicKey pubkey;
     ReturnErrorOnFailure(chip::Crypto::VerifyCertificateSigningRequest(csr.data(), csr.size(), pubkey));
 
+    uint16_t dacVid = VendorId::NotSpecified;
+    uint16_t dacPid = 0;
+    ReturnErrorOnFailure(chip::Crypto::ExtractDNAttributeFromX509Cert(MatterOid::kVendorId, DAC, dacVid));
+    ReturnErrorOnFailure(chip::Crypto::ExtractDNAttributeFromX509Cert(MatterOid::kProductId, DAC, dacPid));
+
+    chip::CATValues assignedCATs;
+    ReturnErrorOnFailure(GetDeviceCATs(static_cast<VendorId>(dacVid), dacPid, assignedCATs));
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:302]: Why are CATs a function of vendor id and product id, exactly?

**Comment by woody-apple** [comment]
@emargolis ?

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm (hunk 17 above)
Comment (emargolis) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:302]: That was my way of thinking that CATs for devices should be a function of `VID` and `PID` values.

For example one of the potential CAT categories will be assigned to all devices that can function as `admins` in the ecosystem. These devices will be recognized by the commissioner using their `PID` values (possibly also `VID`).

Another CAT category can be assigned to all light-switches, which will be also recognized by the commissioner using device `PID`.

File: src/protocols/secure_channel/PASESession.h [hunk 18]
```diff
@@ -276,13 +276,24 @@ class DLL_EXPORT PASESession : public Messaging::ExchangeDelegate, public Pairin
...
 /*
- * The following constants are node IDs that test devices and test
- * controllers use while using the SecurePairingUsingTestSecret to
- * establish secure channel
+ * The following constants are node IDs and CASE Authenticated Tags (CATs)
+ * that test devices and test controllers use while using the
+ * SecurePairingUsingTestSecret to establish secure channel.
  */
 constexpr chip::NodeId kTestControllerNodeId = 112233;
 constexpr chip::NodeId kTestDeviceNodeId     = 12344321;
```
Comment (msandstedt) [inline src/protocols/secure_channel/PASESession.h:284]: This file isn't really a good place for kTestControllerNodeId and kTestDeviceNodeId in the first place.  So adding all of the other test values here seems like we are just going further down the wrong path.  Shouldn't these be in a separate file?

None of these test values would ever be needed in a production build.

File: examples/chip-tool/config/PersistentStorage.cpp [hunk 19]
```diff
@@ -239,3 +240,29 @@ CHIP_ERROR PersistentStorage::SetLocalNodeId(NodeId value)
     uint64_t nodeId = Encoding::LittleEndian::HostSwap64(value);
     return SyncSetKeyValue(kLocalNodeIdKey, &nodeId, sizeof(nodeId));
 }
+
+CATValues PersistentStorage::GetLocalCATs()
```
Comment (msandstedt) [inline examples/chip-tool/config/PersistentStorage.cpp:244]: What are local CATs?  CATs will appear in certificates and in ACLs.

I don't see how this fits into that.

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm (hunk 17 above)
Comment (msandstedt) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:290]: Let's level set though.  It is literally impossible for sdk development to predict what logic a vendor might use to assign CATs.

We therefore need to make sure that any logic along those lines only exists in the concrete operational credential delegate examples.  Such logic cannot e.g. exist in the OperationalCredentialsDelegate interface.

It appears this PR is honoring that.  But I think it's important to highlight this.

File: src/darwin/Framework/CHIP/CHIPDeviceController.mm (hunk 14 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/CHIPDeviceController.mm]: Via whatever API the framework will define, but "read some random storage key" is not necessarily that API.

Please just leave the Darwin code alone and file an issue about adding CAT support there as needed.

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm [hunk 20]
```diff
@@ -100,6 +101,12 @@ static BOOL isRunningTests(void)
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR CHIPOperationalCredentialsDelegate::GetDeviceCATs(VendorId vid, uint16_t pid, CATValues & cats)
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:104]: How would a framework consumer actually use this?  The answer is ... they can't.  We should, imo, be honest and just inline this into the callsite, until we figure out what API will actually be useful here for the framework.

Comment (emargolis) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm]: I removed this logic from the framework and will create a separate ticket to define and implement new APIs to set CAT values.

File: examples/chip-tool/config/PersistentStorage.cpp (hunk 19 above)
Comment (emargolis) [reply examples/chip-tool/config/PersistentStorage.cpp]: I renamed this from  `GetLocalCATs()` to `GetCommissionerCATs()`. This refers to the CATs that will be included in the Commissioner OpCert.

File: src/protocols/secure_channel/PASESession.h (hunk 18 above)
Comment (emargolis) [reply src/protocols/secure_channel/PASESession.h]: removed

File: src/darwin/Framework/CHIP/CHIPDeviceController.mm (hunk 14 above)
Comment (emargolis) [reply src/darwin/Framework/CHIP/CHIPDeviceController.mm]: Updated to assign `kUndefinedCATs`. Will create separate ticket to define proper APIs here.

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm (hunk 17 above)
Comment (emargolis) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm]: I'll create a separate ticket to follow up with proper implementation.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]

**Comment by emargolis** [comment]
@msandstedt PTAL

**Review by msandstedt — APPROVED** [review]

**Pull request merged**